staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -128,35 +145,83 @@ type MutableFeatureGate interface {
 	OverrideDefault(name Feature, override bool) error
 }
 
+// MutableVersionedFeatureGate parses and stores flag gates for known features from
+// a string like feature1=true,feature2=false,...
+// And sets the options for emulated version of the featured gate.
+type MutableVersionedFeatureGate interface {
+	MutableFeatureGate
+	// EmulationVersion returns the version the feature gate is set to emulate.
+	// If set, the feature gate would enable/disable features based on
+	// feature availability and pre-release at the emulated version instead of the binary version.
+	EmulationVersion() *version.Version
+	// SetEmulationVersion overrides the emulationVersion of the feature gate.
+	// Otherwise, the emulationVersion will be the same as the binary version.
+	// If set, the feature defaults and availability will be as if the binary is at the emulated version.
+	SetEmulationVersion(emulationVersion *version.Version) error
+	// DeferErrorsToValidation defers the errors of Set function to the Validate() function if true.
+	// This is used when the user wants to set the feature gate flag before the emulationVersion is finalized.
+	// Validate() should aways be called later to check for flag errors if deferErrorsToValidation is true.
+	DeferErrorsToValidation(val bool)
+	// Validate checks if the flag gates are valid at the emulated version.
+	// Should always be called after Set when DeferErrorsToValidation is set to true.
+	Validate() []error
+	// GetAll returns a copy of the map of known feature names to versioned feature specs.
+	GetAllVersioned() map[Feature]VersionedSpecs
+	// AddVersioned adds versioned feature specs to the featureGate.
+	AddVersioned(features map[Feature]VersionedSpecs) error
+}
+
+// MutableVersionedFeatureGateForTests is a feature gate interface that should only be used in tests.
+type MutableVersionedFeatureGateForTests interface {
+	MutableVersionedFeatureGate
+	// Reset sets the enabled and enabledRaw to empty.
+	Reset()
+	// ForceSet sets the feature to value regardless of its default/lock state.
+	ForceSet(feature Feature, value bool) error
+}
+
 // featureGate implements FeatureGate as well as pflag.Value for flag parsing.
 type featureGate struct {
 	featureGateName string
 
-	special map[Feature]func(map[Feature]FeatureSpec, map[Feature]bool, bool)
+	special map[Feature]func(map[Feature]VersionedSpecs, map[Feature]bool, bool, *version.Version)
 
 	// lock guards writes to known, enabled, and reads/writes of closed
 	lock sync.Mutex
 	// known holds a map[Feature]FeatureSpec
 	known atomic.Value
 	// enabled holds a map[Feature]bool
 	enabled atomic.Value
+	// enabledRaw holds a map[string]bool of the parsed flag
```

Comment:
Say a bit more about how enabledRaw differs from enabled here?  It took me a few minutes to unpack in the code exactly what the role each of these fields have..